lucille-core/src/main/java/com/kmwllc/lucille/core/RunnerManager.java

```diff
@@ -1,80 +1,204 @@
-   * Blocks the calling thread until the current lucille run is completed. This method is primarily intended to be used for testing,
-   * but has been made public to facilitate testing in other modules, namely lucille-plugins/lucille-api.
+   * Starts a new lucille run with the given runId and default configuration. Will not start if a
+   * run with the given runId is already in progress.
+   *
+   * @param runId the unique ID for the lucille run
+   * @return true if the run was started successfully; false if a run with the given ID is already
+   *         in progress
+   * @throws ConfigException 
    */
-  public void waitForRunCompletion() throws ExecutionException, InterruptedException {
-    future.get();
+  public synchronized boolean run(String runId) throws ConfigException {
+    Config config = ConfigFactory.load();
+    String configId = createConfig(config);
+    return runWithConfig(runId, configId);
   }
 
-  private CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+  /**
+   * Run with config and an internal runId will be generated
+   * 
+   * @param config
+   * @return
+   * @throws ConfigException 
+   */
+  protected synchronized boolean runWithConfig(Config config) throws ConfigException {
+    String configId = createConfig(config);
+    return runWithConfig(null, configId);
+  }
 
   /**
-   * Main entrypoint for kicking off lucille runs. This method spawns a new thread for lucille to run in, but will not start a new
-   * instance of lucille until the previous one terminates.
+   * Method to create a lucille run with a custom configuration.
    *
-   * @return boolean representing whether the lucille run was initiated or not. This will return false if and only if the lucille
-   * run was skipped due to the previous run still existing.
+   * @param runId the unique ID for the lucille run
+   * @param config the configuration to use for the run
+   * @return true if the run was started successfully; false if a run with the given ID is already
+   *         in progress
    */
-  synchronized public boolean run() {
-    return runWithConfig(ConfigFactory.load());
+  public synchronized String createConfig(Config config) {
+    String configId = UUID.randomUUID().toString();
+    configMap.put(configId, config);
+    return configId;
   }
 
   /**
-   * Internal abstraction used to support testing.
+   * Method to start a lucille run with a custom configuration.
+   *
+   * @param runId the unique ID for the lucille run
+   * @param config the configuration to use for the run
+   * @return true if the run was started successfully; false if a run with the given ID is already
+   *         in progress
+   * @throws Exception 
    */
-  synchronized protected boolean runWithConfig(Config config) {
-    if (isRunning()) {
-      log.warn("Skipping new run; previous lucille run is still in progress.");
+  public synchronized boolean runWithConfig(String inRunId, String configId) throws ConfigException {
```

Comment:
the argument can be called runId and you can reassign it if it is null
there is not a need for a separate inRunId parameter here

Reply:
I can't do this because runId is required to be final in the async body.

Reply:
OK. The variable name "inRunId" tripped me up when I first saw it. How about "providedRunId" ?

Reply:
Let's require that the runId argument be non-null here. Throw an exception if it's null.

In `runWithConfig(Config config)` where we're currently delegating to runWithConfig(null, configId), we can generate the runId there instead of passing null.

To standardize the way runIds are generated, let's add this to Runner:

```
  /**
   * Utility method to generate a Run ID.
   */
  public static String generateRunId() {
    return UUID.randomUUID().toString();
  }
```
and use it in any place where we need to generate a runId